lib/services/join.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"context"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+// JoinService abstracts the proto.JoinService interface so that it can by
+// implemented by both the auth client and the auth server.
+type JoinService interface {
```

Comment:
I'm not sold on the interface reuse, it seems to go in the opposite direction of usual Go interface usage. I'd be tempted to redefine in the few places where we have it used (here and lib/auth/register.go, I think?).

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
done